manim/renderer/shader.py

```diff
@@ -0,0 +1,272 @@
+
+    def match_updaters(self, mobject):
+        self.clear_updaters()
+        for updater in mobject.get_updaters():
+            self.add_updater(updater)
+        return self
+
+    def suspend_updating(self):
+        self.updating_suspended = True
+        return self
+
+    def resume_updating(self, call_updater=True):
+        self.updating_suspended = False
+        if call_updater:
+            self.update(dt=0)
+        return self
+
+    def refresh_has_updater_status(self):
+        self.has_updaters = len(self.get_updaters()) > 0
+        return self
+
+
+class Shader:
+    def __init__(
+        self,
+        context,
+        name=None,
+        source=None,
+    ):
+        self.context = context
+        self.name = name
+
+        # See if the program is cached.
+        if self.name in shader_program_cache:
+            self.shader_program = shader_program_cache[self.name]
+
+        # Generate the shader from inline code if it was passed.
+        if source is not None:
+            self.shader_program = context.program(**source)
+        else:
+            # Search for a file containing the shader.
+            source_dict = {}
+            source_dict_key = {
+                "vert": "vertex_shader",
+                "frag": "fragment_shader",
+                "geom": "geometry_shader",
+            }
+            shader_folder = SHADER_FOLDER / name
+            for shader_file in os.listdir(shader_folder):
+                shader_file_path = shader_folder / shader_file
+                shader_source = get_shader_code_from_file(shader_file_path)
+                source_dict[source_dict_key[shader_file_path.stem]] = shader_source
+            self.shader_program = context.program(**source_dict)
+
+        # Cache the shader.
+        if name is not None:
+            shader_program_cache[self.name] = self.shader_program
+
+    def set_uniform(self, name, value):
+        self.shader_program[name] = value
+
+
+class FullScreenQuad(Mesh):
+    def __init__(
+        self,
+        context,
+        fragment_shader_source=None,
+        fragment_shader_name=None,
+        output_color_variable="frag_color",
+    ):
+        if fragment_shader_source is None and fragment_shader_name is None:
+            raise Exception("Must either pass shader name or shader source.")
+
+        if fragment_shader_name is not None:
+            # Use the name.
+            shader_file_path = SHADER_FOLDER / f"{fragment_shader_name}.frag"
+            fragment_shader_source = get_shader_code_from_file(shader_file_path)
+        elif fragment_shader_source is not None:
+            # Find the start of the shader.
+            fragment_shader_lines = fragment_shader_source.split("\n")
+            first_non_whitespace_line_index = 0
+            while re.match(
+                r"^\s*$",
+                fragment_shader_lines[first_non_whitespace_line_index],
+            ):
+                first_non_whitespace_line_index += 1
```

Comment:
How about simply `.lstrip()` the `fragment_shader_source` input rather than doing this?

Reply:
Done.